### PR TITLE
Split the debug_line section in MCCAS

### PR DESCRIPTION
**Pull request by rastogishubham**
Use MCCAS to split the debug_line section by adding one block per line table entry. I created my own block kind call DebugLineRef which represents one line table entry

### CODE REVIEW

**Comment by rastogishubham** [comment]
@swift-ci please test

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.h
Comment (adrian-prantl) [inline llvm/include/llvm/MC/CAS/MCCASObjectV1.h]: I think it would be nice to add a Doxygen comment for the function.
Should we call it `createDebugLineSection` for more clarity?

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (adrian-prantl) [inline llvm/lib/MC/MCCASObjectV1.cpp]: why not
```
do {
} while (!SectionReader.empty());
return Error::success();
```
?

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (adrian-prantl) [inline llvm/lib/MC/MCCASObjectV1.cpp]: A short comment explaining what this loop is doing might be nice. Something like:
`// Iterate over the individual .debug_line contributions in this section and pass them to CreatFn.`
and then:
`// Extract the length of this .debug_line contribution from its header.`

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (adrian-prantl) [inline llvm/lib/MC/MCCASObjectV1.cpp]: This could also use a high-level comment.

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (adrian-prantl) [inline llvm/lib/MC/MCCASObjectV1.cpp]: Why the SmallVector and not a vector, and is the default size reasonable, or should we reserve more space for the inline elements?

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (adrian-prantl) [inline llvm/lib/MC/MCCASObjectV1.cpp]: Comment what this condition means?

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (adrian-prantl) [inline llvm/lib/MC/MCCASObjectV1.cpp]: what happens if it's not in RelMap, is that handled below, or an error?

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (adrian-prantl) [inline llvm/lib/MC/MCCASObjectV1.cpp]: is the else path an error? What does it mean if we go there?

**Review by adrian-prantl — COMMENTED** [review]
Nice! I left a few minor, mostly stylistic comments inside.

File: clang/test/CAS/debug-line-split.c
Comment (cachemeifyoucan) [inline clang/test/CAS/debug-line-split.c]: This is test is basically the same as run the clang once with `-Xclang -fcas-backend-mode=verify`.

I am hoping for a better test, e.g., run cas-backend, then dwarfdump the object file to make sure the debug line is correct.

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.h
Comment (cachemeifyoucan) [reply llvm/include/llvm/MC/CAS/MCCASObjectV1.h]: Agree.

Also it might just be better the CreateFn is `Error(ArrayRef<char>)`? You can slice ArrayRef and don't need that StringRef conversion?

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (cachemeifyoucan) [reply llvm/lib/MC/MCCASObjectV1.cpp]: I will suggest move this to the function where `RelMap` is built. You can skip building the map in DebugLine section and directly add to SectionRelocs because this section is encoded differently without considering Atoms.

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (cachemeifyoucan) [reply llvm/lib/MC/MCCASObjectV1.cpp]: Agree. This should be an assert or unreachable.

**Review by cachemeifyoucan — COMMENTED** [review]
Some feedback inline.

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [reply llvm/lib/MC/MCCASObjectV1.cpp]: In general, the LLVM's programmer manual prescribes [always favouring](https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h) `SmallVector<T, 0>` over `std::vector`.

> SmallVector has grown a few other minor advantages over std::vector, causing SmallVector<Type, 0> to be preferred over std::vector<Type>. [...]

Regarding the "SmallSize" default, do we have any good metrics for what a type DebugLine section would look like for a "Small" function?

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: Does this section have more than one fragment? IIRC the "debug info" section only has a single fragment

Edit: NVM, I see the dyn_casts below which indicate multiple fragments

**Review by felipepiovezan — COMMENTED** [review]
This will have a lot overlap with the work for the debug section, so I'll wait for this patch to be merged before posting my PR; some of the comments/resolutions here will also apply to the debug info section.

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.h
Comment (felipepiovezan) [reply llvm/include/llvm/MC/CAS/MCCASObjectV1.h]: I think the challenge is that we cannot create a `BinaryStreamReader` with an `ArrayRef<char>`, only `ArrayRef<uint8_t>` and `StringRef`

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: Should we make this a helper function? I'll need it too. This is what I have right now:

```cpp
/// Reads the Length field of the debug info header contained in Reader,
/// assuming Reader is positioned at the beginning of the section.
/// The reader state is advanced by however many bytes are used by the Length
/// field, plus Length bytes.
Expected<size_t> getSizeFromDebugInfoHeaderAndSkip(BinaryStreamReader &Reader) {
  // From Dwarf 5 section 7.4:
  // In the 32-bit DWARF format, an initial length field [...] is an unsigned
  // 4-byte integer (which must be less than 0xfffffff0);
  uint32_t Word1;
  if (auto E = Reader.readInteger(Word1))
    return E;

  // TODO: handle 64-bit DWARF format.
  assert(Word1 < 0xfffffff0);

  if (auto E = Reader.skip(Word1))
    return E;
  return Word1;
}
```


And then later, I use this function like this:

```cpp
  while (!Reader.empty()) {
    auto StartOffset = Reader.getOffset();
    auto Length = getSizeFromDebugInfoHeaderAndSkip(Reader);
    if (!Length)
      return Length.takeError();

    // TODO: zero out abbrev offset once it's been split
    auto SectionData =
        DebugInfoStrRef.substr(StartOffset, Reader.getOffset() - StartOffset);
    auto DbgInfoRef = DebugInfoRef::create(*this, SectionData);

```

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (adrian-prantl) [reply llvm/lib/MC/MCCASObjectV1.cpp]: > In general, the LLVM's programmer manual prescribes [always favouring](https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h) SmallVector<T, 0> over std::vector.

TIL!

Comment (adrian-prantl) [reply llvm/lib/MC/MCCASObjectV1.cpp]: > Regarding the "SmallSize" default, do we have any good metrics for what a type DebugLine section would look like for a "Small" function?

This is probably a microoptimization that is largely irrelevant, but we could gather statistics from e.g., a build of Clang if we wanted to. But really, size 0 seems fine, too.

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.h
Comment (rastogishubham) [reply llvm/include/llvm/MC/CAS/MCCASObjectV1.h]: Yep, I need the `StringRef` to convert it to a `BinaryStreamReader` which will take care of slicing with endianness in mind

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: Yes that can be done, I will change it

File: clang/test/CAS/debug-line-split.c
Comment (rastogishubham) [reply clang/test/CAS/debug-line-split.c]: Yep you are completely right!

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: I will be honest, I just used a `SmallVector` because I thought it was more common to use it in LLVM. I have no idea what would be more ideal

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: @felipepiovezan I am pretty sure that the Debug Info section can have multiple fragments, and you should not assume that it will always be one fragment. You could have multiple data fragments as well.

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (adrian-prantl) [reply llvm/lib/MC/MCCASObjectV1.cpp]: Well, from the document Felipe linked, you were right :-)

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: Yes I think this is a better approach and I will move the code there, thanks Steven!

**Comment by rastogishubham** [comment]
@swift-ci please test

**Comment by rastogishubham** [comment]
@cachemeifyoucan I changed the test a little to check that two functions indeed produce two blocks in the object-stats

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: `assert(false, msg)` => `llvm_unreachable(msg)`

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: I understand the desire to account for the 64 bit format, but right now this code is both unreachable and untestable, so IMO we should remove it

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: As discussed in the other PR, this is not an "if and only if statement".
The 64-bit format will use `0xffffffff`, but this doesn't imply that the 32-bit format uses _anything_ different from `0xffffffff`

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: I understand the debug str handling has been inlined in the middle of the loop, but it likely won't scale very well once all debug sections are accounted for. As such we could outline the body of the `if` into a helper function.

**Comment by rastogishubham** [comment]
@swift-ci please test

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: We're calling this in a loop, we should hoist this in case it's not a cheap computation

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp:1454]: @rastogishubham have you actually _seen_ any DWARF debug info sections with multiple fragments? Nowhere in the code I see how the original fragments are materialized in their original state, which makes me believe we're not testing the multiple-fragment case.

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: Good catch

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: I don't think it is hurting anything and it is correct, it is not like this will change. So why rewrite this in the future?

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: Yep, I hadn't realized that, I will make the change

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp:1454]: If you step through the code for Debug Line, you will see that there are at least 10 - 12 fragments for the debug line section, in a small contrived test like 

int bar() {
   return 1;
   }
int foo() {
 return 2;
}
int baz() {
  return 3;
}
So yes, I have seen multiple fragments, and I think we shouldn't assume that there will always be one unless we find solid proof in the code that there is only one

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: The only reason I wrote it inline was that this function will only be called once, but I do agree with you, it doesn't scale well for every debug information section. I will make a separate function for it, good catch!

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: That is fair, though I would have imagined that the compiler might optimize for that? I will take it out of the loop regardless, good catch!

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [reply llvm/lib/MC/MCCASObjectV1.cpp]: > it is correct

I will defer to others here, but IMO code that is not tested cannot be said to be correct

Comment (adrian-prantl) [reply llvm/lib/MC/MCCASObjectV1.cpp]: > but IMO code that is not tested cannot be said to be correct

Yeah, let's either stick an llvm_unreachable there, or find a way to test it.

Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: It has been changed with the rebase :)

**Comment by rastogishubham** [comment]
@felipepiovezan This is what it looks like after the rebase, I wanted to get a consensus on what to name the individual splitting functions, since the debug_string splitter is just called `createStringSection`, do you just want to stay with the same convention and change the name to from `splitDebugInfoCUs` to `createInfoSection`?

@adrian-prantl do you have any other comments?

**Comment by rastogishubham** [comment]
@swift-ci please test

**Comment by rastogishubham** [comment]
@swift-ci please test

**Comment by rastogishubham** [comment]
@swift-ci please test

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: Why did we change this from a dyn_cast to a "Kind" test followed by a cast? The dyn_cast approach seems more idiomatic to what the LLVM codebase uses. 

I could be missing some details here, but this change seems unrelated to what the PR is trying to accomplish.

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: In general, it is better to reduce the diff to present reviewers with a minimal patch, I'm a bit confused as to what caused the renaming

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: This change also seems unrelated to the PR?

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: Small nit, but we don't need to repeat the type twice here:

`StringRef DebugLineStrRef(DebugLineData->data(), DebugLineData->size());`

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: I believe all of these comments would be better as the documentation of what `createLineSection` does, not as inline comments. Note how the current documentation of `createLineSection` is much simpler than the description above.

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: I think the comment is not reflecting the line below, as it only mentions the Line sections, but the line below refers to both debug info and debug section.

I think we could either remove the comment or make it more generic and just say "Dwarf Section"

**Review by felipepiovezan — COMMENTED** [review]
This LGTM in general, but some pieces of the PR confused me as they seem unrelated to what the PR is trying to do.

What is our plan with respect to the commits? It would be nice to squash them and present the reviewers with what the final proposed patch is going to look like

**Comment by felipepiovezan** [comment]
> do you just want to stay with the same convention and change the name to from `splitDebugInfoCUs` to `createInfoSection`?

I don't have any preferences here, but we could address these differences on a separate NFC commit

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: Two reasons:

1. Look at the if statement below, I need to also check to make sure there is an `MCDwarfLineAddrFragment` only if the section is a debug_line section. 

2. Imo, a check for the `Kind` is more efficient than the cast

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: When I rebased, I wanted to make things consistent, so I just added the MCSection as a parameter. If you think we should have a separate PR just to make things consistent, that is fine too, but I didn't see it that way

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: Yep! Good catch!

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: Yep! Good catch!

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: I see, so you are saying I should move this comment to the header file? That works too!

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: This is my bad, I had a similar function and when resolving the merge conflict I just went with F instead of Fragment

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [reply llvm/lib/MC/MCCASObjectV1.cpp]: > Imo, a check for the Kind is more efficient than the cast

They are virtually the same thing for builds with assertions. So the current form is doing the same check twice (a `cast` will check too).

> need to also check to make sure there is an MCDwarfLineAddrFragment

Wouldn't this do the job?

```cpp
else if (const auto *LineFragment = dyn_cast<MCDwarfLineAddrFragment>(&Fragment))
  if(IsDebugLineSection)
    llvm::append_range(...)
```

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [reply llvm/lib/MC/MCCASObjectV1.cpp]: In general I'd just favour minimizing the diff (makes it easier for reviwers) and keeping NFC changes separate

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [reply llvm/lib/MC/MCCASObjectV1.cpp]: Yup, I think it would be easier to find the information that way

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: It would, but when I wrote I just went a different way, especially because if you see any of the code in `MCAssembler.cpp` anything to do with the fragment switches on the fragment kind so I want to keep that consistent. If it could all be achieved with a cast, then why add a Kind to the class anyway? If the dyn_cast is more idiomatic to the llvm coding style, then I do not know why MCAssembler uses the Kind in a switch.

Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: I changed it to a dyn_cast :)

**Comment by rastogishubham** [comment]
@swift-ci please test

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [reply llvm/lib/MC/MCCASObjectV1.cpp]: Awesome, thank you!

> why MCAssembler uses the Kind in a switch.

Because they need a switch ;)
C++ doesn't allow switching on a type, more recently LLVM gained the ability to do so with some template tricks. See for example https://llvm.org/doxygen/TypeSwitch_8h_source.html#l00107

**Review by felipepiovezan — APPROVED** [review]
LGTM! Thanks for all the changes

**Review by ghost — APPROVED** [review]

**Pull request merged**
